snuba/query/project_extension.py

```diff
@@ -0,0 +1,89 @@
+from snuba import settings, util
+from snuba.state import get_config
+from snuba.replacer import get_projects_query_flags
+from snuba.query.extensions import ExtensionQueryProcessor, QueryExtension
+from snuba.query.query_processor import ExtensionData
+from snuba.query.query import Query, QueryHints
+
+
+PROJECT_EXTENSION_SCHEMA = {
+    'type': 'object',
+    'properties': {
+        'project': {
+            'anyOf': [
+                {'type': 'integer', 'minimum': 1},
+                {
+                    'type': 'array',
+                    'items': {'type': 'integer', 'minimum': 1},
+                    'minItems': 1,
+                },
+            ]
+        },
+    },
+    # Need to select down to the project level for customer isolation and performance
+    'required': ['project'],
+    'additionalProperties': False,
+}
+
+
+class ProjectExtensionProcessor(ExtensionQueryProcessor):
+
+    def process_query(
+            self,
+            query: Query,
+            extension_data: ExtensionData,
+            query_hints: QueryHints,
+    ) -> None:
+        # NOTE: we rely entirely on the schema to make sure that regular snuba
+        # queries are required to send a project_id filter. Some other special
+        # internal query types do not require a project_id filter.
+        project_ids = util.to_list(extension_data['project'])
+
+        if project_ids:
+            query.add_conditions([('project_id', 'IN', project_ids)])
+
+
+class ProjectWithGroupsProcessor(ProjectExtensionProcessor):
+
+    def process_query(
+            self,
+            query: Query,
+            extension_data: ExtensionData,
+            query_hints: QueryHints,
+    ) -> None:
+        # NOTE: we rely entirely on the schema to make sure that regular snuba
```

Comment:
I don't think this comment is relevant anymore since we have datasets without project extension